README.md
# django-scaffold

The objective of this project is to provide a quick scaffold for a django project that can be used as a reference for your own project.

[![Pre-commit](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/pre-commit.yml/badge.svg)](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/pre-commit.yml)
[![Linting](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/lint.yml/badge.svg)](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/lint.yml)
[![Testing](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/test.yml/badge.svg)](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/test.yml)
[![Documentation](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/docs-pages.yml/badge.svg)](https://github.com/zen-code-symphony/django-scaffold/actions/workflows/docs-pages.yml)

# Table of Contents
- [Prerequisites](#prerequisites)
  - [Install software](#install-software)
- [Get up and running](#get-up-and-running)
  - [PATH 1: Create a copy for your own project](#path-1-create-a-copy-for-your-own-project)
  - [PATH 2: Cloning this repository](#path-2-cloning-this-repository)
  - [Build and run app](#build-and-run-app)
- [Configure VS Code](#configure-vs-code)
- [Learning resources](#learning-resources)


## Prerequisites

### Install software
- [docker](https://docs.docker.com/get-docker/)
- [git](https://git-scm.com/downloads)
- [pyenv - _optional_](https://github.com/pyenv/pyenv)
- [VS Code - _optional_](https://code.visualstudio.com/download)


## Get up and running

Follow one of the below paths to get up and running. **NOTE**: It assumes that you have installed Python and configured git. If you are using pyenv, `.python-version` file will be used.

### PATH 1: Create a copy for your own project

Run the below command to create a project based on this basic scaffold.

**Option 1**: Create new project from Linux or macOS command shell and stay in the shell:
```sh
# Replace "my_project" with your own project name.
curl -sSL https://raw.githubusercontent.com/zen-code-symphony/django-scaffold/main/create-project.sh | bash -s my_project && cd my_project && source app/venv/bin/activate
```

**Option 2**: Create new project from Linux or macOS shell and open VS Code editor:
```sh
# Replace "my_project" with your own project name.
curl -sSL https://raw.githubusercontent.com/zen-code-symphony/django-scaffold/main/create-project.sh | bash -s my_project && cd my_project && code .
```

### PATH 2: Cloning this repository

1. Clone the repository.
```sh
git clone git@github.com:zen-code-symphony/django-scaffold.git && cd django-scaffold
```
2. Create virtual environment, activate it, install dependencies, and setup pre-commit hooks.
```sh
make init
```

### Build and run app
Once you have followed [PATH 1](#path-1-create-a-copy-for-your-own-project) or [PATH 2](#path-2-cloning-this-repository) above, build and run the app in dev (*Option 1* below) or prod (*Option 2* below) mode.
* **[Option 1]** Run in development mode by following below instructions:
  * 1.1 Build and run the dev container image(s).
  ```sh
  make bup-dev
  ```
  * 1.2 Browser will open automatically with the app running at [http://localhost:8000/](http://localhost:8000/)
* **[Option 2]** Run in production mode by following below instructions:
    * 2.1. Build and run the prod container image(s).
    ```sh
    make bup-prod
    ```
    * 2.2. Browser will open automatically with the app running at [http://localhost:1337/](http://localhost:1337/)


## Configure VS Code
  - As part of VS Code workspace extensions recommendations, the extensions mentioned in [.vscode/extensions.json](./.vscode/extensions.json), should be installed. This includes extensions for flake8, black, isort etc.
  - Make sure that you are using the workspace settings mentioned in repository file [.vscode/settings.json](./.vscode/settings.json). `CTRL+,` and open `Workspace` settings tab to check.

## References
* [Black](https://black.readthedocs.io/en/stable/index.html)
* [Conventional Commits](https://www.conventionalcommits.org/)
* [djangoproject.com](https://www.djangoproject.com/)
* [diataxis](https://diataxis.fr/)
* [docs.docker.com](https://docs.docker.com/)
* Flake8
  - [Flake8](https://flake8.pycqa.org/en/latest/index.html)
  - [VSCode-Flake8](https://github.com/microsoft/vscode-flake8)
* [Git](https://git-scm.com/book/en/v2)
* [GitHub Actions](https://docs.github.com/en/actions)
* [Markdown cheat-sheet](https://www.markdownguide.org/cheat-sheet/)
* [MkDocs](https://www.mkdocs.org/)
* [mypy](https://mypy.readthedocs.io/en/stable/index.html)
* [pip-tools](https://pip-tools.readthedocs.io/en/latest/)
* [Python best practices for a new project ](https://mitelman.engineering/blog/python-best-practice/automating-python-best-practices-for-a-new-project/)
* [realpython.com](https://realpython.com/)
* [Semantic Versioning](https://semver.org/)
* [Setting up Python projects](https://towardsdatascience.com/setting-up-python-projects-part-i-408603868c08)
* [stackoverflow.com](https://stackoverflow.com/)
* [testdriven.io](https://testdriven.io/)


CHANGELOG.md
# CHANGELOG



## v0.0.0 (2023-12-30)

### Build

* build: Added version number to project. ([`ee76f19`](https://github.com/zen-code-symphony/django-scaffold/commit/ee76f19b35162d612250f8d8dcd8abbd070a3c2d))

### Chore

* chore: Added target for mkdocs serve. ([`44b39f8`](https://github.com/zen-code-symphony/django-scaffold/commit/44b39f83793dfc50003005ff5ac7567c70fb8d61))

* chore: Updated Makefile. ([`d8a22c8`](https://github.com/zen-code-symphony/django-scaffold/commit/d8a22c866caf627accc920465cb2c9bd87d4c294))

* chore: Updated Makefile. ([`4faca18`](https://github.com/zen-code-symphony/django-scaffold/commit/4faca18e5f9f9588e33821570cdf3fbc22a6ebe9))

* chore: Updated docs target as .PHONY ([`49e3aaf`](https://github.com/zen-code-symphony/django-scaffold/commit/49e3aafb33cbc5f9b5ee52e1302c9dc5da8320cd))

* chore: Updated Makefile to include docs targets. ([`4c2b365`](https://github.com/zen-code-symphony/django-scaffold/commit/4c2b3651f2391da0e2de128970a7cffe589a6b0f))

* chore(pre-commit): Added conventional commit pre-commit hook. ([`cfd0a08`](https://github.com/zen-code-symphony/django-scaffold/commit/cfd0a0874150b9deffe883d09bf78a8d49926bde))

### Ci

* ci: Added workflow for documentation generation. ([`72d503d`](https://github.com/zen-code-symphony/django-scaffold/commit/72d503dbb1bc3ebd7af0241add88d998cb89a9e1))

### Documentation

* docs: Added autogeneration of code documentation using mkdocstrings. ([`8873d94`](https://github.com/zen-code-symphony/django-scaffold/commit/8873d9481e5ada3cb3ada51c9ccf110481b3a685))

* docs: Updated Makefile and docs. ([`46db5a3`](https://github.com/zen-code-symphony/django-scaffold/commit/46db5a33ffdff768cccc1bab34825f0bcedd4673))

* docs: Added WIP note to the docs. ([`6aa5c2d`](https://github.com/zen-code-symphony/django-scaffold/commit/6aa5c2d5254276f537da37bf9086f26523d646c7))

* docs: Removed unwanted comment. ([`038a570`](https://github.com/zen-code-symphony/django-scaffold/commit/038a57085c31034ae2c940b0cff6679b197c2395))

* docs: Added WIP note to MkDocs. ([`6d50f35`](https://github.com/zen-code-symphony/django-scaffold/commit/6d50f3536146b12b8782ebed83f872af5f7acf91))

* docs: Added repository URL to the mkdocs.yml. ([`1c26a1c`](https://github.com/zen-code-symphony/django-scaffold/commit/1c26a1ca6cc84a06f44730c473c983ff2e9ec4c9))

* docs: Updated references. ([`1f599dc`](https://github.com/zen-code-symphony/django-scaffold/commit/1f599dc29798b4db01fd629921c96f6b98f35fe3))

* docs: Updated Makefile with new update and docs targets. ([`5380444`](https://github.com/zen-code-symphony/django-scaffold/commit/53804448b191e6e1478beb026504854f4c738195))

* docs: Added MkDocs template based on Diátaxis framework. ([`2342734`](https://github.com/zen-code-symphony/django-scaffold/commit/2342734cd5f58c2463e6206bda727dbaf7d9d518))

* docs: Updated links in learning resources. ([`dea0d96`](https://github.com/zen-code-symphony/django-scaffold/commit/dea0d96874f6b2c0e542fa464cc581fe431d6e6b))

### Unknown

* Updated gitignore and Makefile clean target. ([`4077e84`](https://github.com/zen-code-symphony/django-scaffold/commit/4077e84a20d84fcfe8976e9bbb62f3d7e0fd0615))

* Updated .dockerignore file. ([`b9a2a71`](https://github.com/zen-code-symphony/django-scaffold/commit/b9a2a712620b2cdab51da0ea4465830a512c5204))

* Added mypy extension to recommendations and added settings for the extension. ([`484c9d6`](https://github.com/zen-code-symphony/django-scaffold/commit/484c9d655299a27c23724c39fe2f2fb9df046f7c))

* Updated README.md for VS Code configuration. ([`f103857`](https://github.com/zen-code-symphony/django-scaffold/commit/f10385729b2fa4df902640079228e0c27b5642ba))

* Merge pull request #3 from zen-code-symphony/chore/ga-workflow

Added coverage report in workflow summary. ([`7f137a9`](https://github.com/zen-code-symphony/django-scaffold/commit/7f137a9a6fe8c57462beee5f3950223f3051710f))

* Added coverage report in workflow summary. ([`3eb0bff`](https://github.com/zen-code-symphony/django-scaffold/commit/3eb0bff80810fc5479122c83bc9f34ee1a20890a))

* Update README.md ([`e68ffcd`](https://github.com/zen-code-symphony/django-scaffold/commit/e68ffcd56a6a1f7e4b6a5cb15d0d022fa3b6b240))

* Merge pull request #2 from zen-code-symphony/chore/ga-workflow

Refactored workflow to pre-commit and updated test workflow to run manage.py tests as well ([`dbd205e`](https://github.com/zen-code-symphony/django-scaffold/commit/dbd205e5b396c8b83916c2acd1f56d8faefd1794))

* Refactored workflow to pre-commit and updated test workflow to run manage.py tests as well. ([`7aa3e6d`](https://github.com/zen-code-symphony/django-scaffold/commit/7aa3e6d713bd2d06b80445b29f80faac76f176b9))

* Update README.md ([`072c56f`](https://github.com/zen-code-symphony/django-scaffold/commit/072c56f8cd64ce8a03b07e09345ea9f83ff3b82f))

* 1. Added pytest, pytest-django, pytest-cov.
2. Added make targets to run tests.
3. Added GA workflow for tests. ([`e1ae3bc`](https://github.com/zen-code-symphony/django-scaffold/commit/e1ae3bcd92c0f4a178b6a2fd7ef0146bbd78ec7d))

* Added formating and help doc for make targets. ([`1abcd07`](https://github.com/zen-code-symphony/django-scaffold/commit/1abcd07eac0a096917e111245889cbd955bfd26a))

* Update README.md ([`42cc42d`](https://github.com/zen-code-symphony/django-scaffold/commit/42cc42deef7280f20736dbb7abb40401aca4bc63))

* Merge pull request #1 from zen-code-symphony/feat/lint-ci

Added lint CI ([`9acbb0e`](https://github.com/zen-code-symphony/django-scaffold/commit/9acbb0e962c983c77080f59845812798a35bdf04))

* Installing dev requirements for linting tools. ([`3671a52`](https://github.com/zen-code-symphony/django-scaffold/commit/3671a5236e7ece61a2f7a967d1d79cddafda4355))

* Added lint CI ([`c6f87cf`](https://github.com/zen-code-symphony/django-scaffold/commit/c6f87cf0452fb9f47c263d4871acf432b2a16f46))

* Update README.md

Added badge for workflow. ([`7d13f69`](https://github.com/zen-code-symphony/django-scaffold/commit/7d13f69609114aaa38cdaa6a61440a24ecbcee03))

app/project/urls.py
"""
URL configuration for project project.

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/5.0/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""

# from django.conf import settings
# from django.conf.urls.static import static
from django.contrib import admin
from django.urls import path

urlpatterns = [
    path("admin/", admin.site.urls),
]

# Enable this once you add app. Keeping it disabled to
# keep the default Django index page active.
# if bool(settings.DEBUG):
#     urlpatterns += static(
#         settings.MEDIA_URL, document_root=settings.MEDIA_ROOT
#     )


app/project/settings.py
"""
Django settings for project project.

Generated by 'django-admin startproject' using Django 5.0.

For more information on this file, see
https://docs.djangoproject.com/en/5.0/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/5.0/ref/settings/
"""

import os
from pathlib import Path

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/5.0/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get("SECRET_KEY", "")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = bool(int(os.environ.get("DEBUG", default=0)))

ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS", "localhost").split(" ")


# Application definition

INSTALLED_APPS = [
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "project.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "project.wsgi.application"


# Database
# https://docs.djangoproject.com/en/5.0/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": os.environ.get("SQL_ENGINE", "django.db.backends.sqlite3"),
        "NAME": os.environ.get("SQL_DATABASE", BASE_DIR / "db.sqlite3"),
        "USER": os.environ.get("SQL_USER", "user"),
        "PASSWORD": os.environ.get("SQL_PASSWORD", "password"),
        "HOST": os.environ.get("SQL_HOST", "localhost"),
        "PORT": os.environ.get("SQL_PORT", "5432"),
    }
}


# Password validation
# https://docs.djangoproject.com/en/5.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",  # noqa: 501
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",  # noqa: 501
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",  # noqa: 501
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",  # noqa: 501
    },
]


# Internationalization
# https://docs.djangoproject.com/en/5.0/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/5.0/howto/static-files/

STATIC_URL = "static/"
STATIC_ROOT = BASE_DIR / "staticfiles"

# Media files (uploads)
MEDIA_URL = "/media/"
MEDIA_ROOT = BASE_DIR / "mediafiles"

# Default primary key field type
# https://docs.djangoproject.com/en/5.0/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

# A list of trusted origins for unsafe requests (e.g. POST).
CSRF_TRUSTED_ORIGINS = ["http://localhost:1337"]
